### PR TITLE
Structural Refactoring - .NET 6.0 & SqlServer

### LINKED ISSUE #40: ✨ [FEATURE] SQL Server Implementation
## Summary
`IronHook` must be support SQL Server.

There should be a library named IronHook.SqlServer

## Advantage
Opening up and increasing usage of IronHook to wider audiences

### LINKED ISSUE #44: 📌 [CUSTOM] Net6 support
## Summary

`IronHook` must be support Net6.

Multiple target framework feature should be used for this operation.

Both Net5 and Net6 should be supported.

### PR DESCRIPTION
**Pull request by enisn**
## Summary
Why do you need to create DbContext and models for each EF Provider? They're just providers, the project abstraction shouldn't know anything about a provider. So I removed all the provider-specified codes and merged them into a single abstraction layer.

As I can see, migrations are managed by the library. Between EF Providers, the only different thing is migrations. Only migrations are separated in the ideal scenario. You can see, provider specified projects have only migrations but now any of EF providers can be used via generating migrations

---

## What brings that structure?
That structure has a couple of benefits like I presented below.

### Database Independency
Firstly, database dependency is completely removed. That means developers can use any database as they wish with EntityFramework. They need to manage their own migrations for database providers which we don't provide migrations.

- New usage of a database that ironhook support (PostgreSql).
If we provide migrations, it can be defined via calling `UseIronHook{databasename}Migrations()` method.
```csharp
            services.AddIronHook(options =>
            {
                options.UseNpgsql(
                Configuration.GetConnectionString("Default"), 
                opts => opts.UseIronHookNpgsqlMigrations());
            });
```
I just replaced `UseIronHook()` method with `MigrateIronHook()` method since it doesn't add a middleware, it just performs a migration operation.

```csharp
app.MigrateIronHook();
```

- New usage of Non-supported databases

```csharp
options.UseMySql(Configuration.GetConnectionString("Default"));
```

Migrations aren't provided by IronHook, but `MigrateIronHook` can still be used, it'll create a database from the current state of DbContext with https://github.com/FowApps/IronHook/pull/45/files#diff-48af72c016bfcfaede0f855ae54a6f1b88fceeb48b7977aea5bb3bc412f376edR41-R48

```csharp
app.MigrateIronHook();
```
> Managing own migrations by developers is highly recommended for unsupported databases but it's not mandatory

### Multi-Targeting
Now, the project can support both framework versions at the same time. Projects are targeting .NET 5 and .NET6 right now.

In the future, .Net 5 support can be dropped partially.

### Common DbContext
I just have removed DB provider-specific DbContexes since there is no difference with CoreDbContext. Using single DbContext is much more maintainable.

---

- Closes #40 
- Closes #44

### CODE REVIEW

**Comment by enisn** [comment]
Hi @codeclimate, all of problems you found are auto-generated files (migrations) 😅

File: src/IronHook.EntityFrameworkCore.SqlServer/IronHookCoreDbContextFactory.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿using IronHook.EntityFrameworkCore.SqlServer.Extensions;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Design;
+
+namespace IronHook.EntityFrameworkCore.SqlServer
+{
+    public class IronHookCoreDbContextFactory : IDesignTimeDbContextFactory<IronHookCoreDbContext>
```
[the remark this answers is not in the document]
Comment (enisn) [reply src/IronHook.EntityFrameworkCore.SqlServer/IronHookCoreDbContextFactory.cs:7]: Nope, only names are same. They were created for totally different purposes for different providers

**Comment by furkandeveloper** [comment]
Thanks for the contribution, @enisn  we are reviewing it as a team.

**Comment by furkandeveloper** [comment]
Update codeclimate.yml file on root and exclude SqlServer migration files. @enisn

**Comment by furkandeveloper** [comment]
Since dotnet6 support has been added, there should be a sample project for dotnet6 in the sample folder. @enisn

**Comment by furkandeveloper** [comment]
Codeclimate treats files with the same name as duplications. For this reason, we must make the names of the IronHookCoreDbContextFactory classes provider dependent.

for example;

IronHookPostgreSqlCoreDbContextFactory instead of IronHookCoreDbContextFactory
we can use

@enisn

**Comment by enisn** [comment]
Any updates?

**Comment by furkandeveloper** [comment]
> Any updates?

Not for now. Tests continue. PR can be merged on weekdays. @enisn

**Review by furkandeveloper — APPROVED** [review]
:lgtm:

**Comment by enisn** [comment]
I can't update the [wiki](https://github.com/FowApps/IronHook/wiki) section, Can you update documentation also?

**Comment by furkandeveloper** [comment]
I updated wiki documentation. Thanks @enisn. Developing with you is so much fun. 🎉🎉🎉🎉

**Comment by furkandeveloper** [comment]
@all-contributors please add @enisn for infrastructure, tests and code

**Pull request merged**